src/SegmentationCameraSensor.cc

```diff
@@ -0,0 +1,511 @@
+  this->dataPtr->camera->LabelMapFromColoredBuffer(
+    this->dataPtr->segmentationLabelsBuffer);
+}
+
+//////////////////////////////////////////////////
+bool SegmentationCameraSensor::Update(
+  const std::chrono::steady_clock::duration &_now)
+{
+  IGN_PROFILE("SegmentationCameraSensor::Update");
+  if (!this->dataPtr->initialized)
+  {
+    ignerr << "Not initialized, update ignored.\n";
+    return false;
+  }
+
+  if (!this->dataPtr->camera)
+  {
+    ignerr << "Camera doesn't exist.\n";
+    return false;
+  }
+
+  // don't render if there are no subscribers
+  if (!this->dataPtr->coloredMapPublisher.HasConnections() &&
+    !this->dataPtr->labelsMapPublisher.HasConnections() &&
+    this->dataPtr->imageEvent.ConnectionCount() <= 0u)
+  {
+    return false;
+  }
+
+  // Actual render
+  this->Render();
+
+  if (!this->dataPtr->segmentationColoredBuffer ||
+    !this->dataPtr->segmentationLabelsBuffer)
+    return false;
+
+  auto width = this->dataPtr->camera->ImageWidth();
+  auto height = this->dataPtr->camera->ImageHeight();
+
+  // create colored map message
+  this->dataPtr->coloredMapMsg.set_width(width);
+  this->dataPtr->coloredMapMsg.set_height(height);
+  // format
+  this->dataPtr->coloredMapMsg.set_step(
+    width * rendering::PixelUtil::BytesPerPixel(rendering::PF_R8G8B8));
+  this->dataPtr->coloredMapMsg.set_pixel_format_type(
+    msgs::PixelFormatType::RGB_INT8);
+  // time stamp
+  auto stamp = this->dataPtr->coloredMapMsg.mutable_header()->mutable_stamp();
+  *stamp = msgs::Convert(_now);
+  auto frame = this->dataPtr->coloredMapMsg.mutable_header()->add_data();
+  frame->set_key("frame_id");
+  frame->add_value(this->Name());
+
+  this->dataPtr->labelsMapMsg.CopyFrom(this->dataPtr->coloredMapMsg);
+
+  // Protect the data being modified by the segmentation buffers
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+
+  // segmentation colored map data
+  this->dataPtr->coloredMapMsg.set_data(
+    this->dataPtr->segmentationColoredBuffer,
+    rendering::PixelUtil::MemorySize(rendering::PF_R8G8B8,
+    width, height));
+
+  // segmentation labels map data
+  this->dataPtr->labelsMapMsg.set_data(this->dataPtr->segmentationLabelsBuffer,
+      rendering::PixelUtil::MemorySize(rendering::PF_R8G8B8,
+      width, height));
+
+  // Publish
+  this->PublishInfo(_now);
+  this->dataPtr->coloredMapPublisher.Publish(this->dataPtr->coloredMapMsg);
+  this->dataPtr->labelsMapPublisher.Publish(this->dataPtr->labelsMapMsg);
+
+  // Trigger callbacks.
+  if (this->dataPtr->imageEvent.ConnectionCount() > 0u)
+  {
+    try
+    {
+      this->dataPtr->imageEvent(this->dataPtr->coloredMapMsg);
+    }
+    catch(...)
+    {
+      ignerr << "Exception thrown in an image callback.\n";
+    }
+  }
+
+  // Save image
+  if (this->dataPtr->saveImage)
+  {
+    this->dataPtr->SaveImage(this->dataPtr->segmentationColoredBuffer,
+      this->dataPtr->segmentationLabelsBuffer, width, height);
+  }
```

Comment:
Would it be better to call this in `OnNewSegmentationFrame`, so that we save images whenever they are received?

Reply:
We don't want to save all images unless the user want to save them, so if the user didn't specify that he wants to save in the SDF, nothing will be saved